### ISSUE TITLE
Compilation problem in the authssh2 policy set

### ISSUE THREAD

**Issue by vmaffione** [IRATI/stack#1096]
This issue probably depends on the openssl version.
I have both openssl 1.0.2.k-3 and 1.1.0.e-1 installed on my systems, but I don't know which one is used for IRATI. Probably the newest.

````
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I/home/vmaffione/git/nxw/irati/librina/src -DSTUB_API=0 -I/home/vmaffione/git/nxw/irati/librina/include -I/home/vmaffione/git/nxw/irati/librina/src/jsoncpp -I/usr/include/libnl3 -I/usr/include/libnl3 -pthread -Wall -Wformat -Wnon-virtual-dtor -Woverloaded-virtual -Werror -Wfatal-errors -g -O0 -fno-inline -g -O2 -MT librina_la-security-manager.lo -MD -MP -MF .deps/librina_la-security-manager.Tpo -c /home/vmaffione/git/nxw/irati/librina/src/security-manager.cc  -fPIC -DPIC -o .libs/librina_la-security-manager.o
/home/vmaffione/git/nxw/irati/librina/src/security-manager.cc: In member function 'void rina::AuthSSH2PolicySet::edh_init_params()':
/home/vmaffione/git/nxw/irati/librina/src/security-manager.cc:778:15: error: invalid use of incomplete type 'DH {aka struct dh_st}'
  dh_parameters->p = BN_bin2bn(dh2048_p,sizeof(dh2048_p),NULL);
               ^~
compilation terminated due to -Wfatal-errors.
make[3]: *** [Makefile:1241: librina_la-security-manager.lo] Error 1
make[3]: Leaving directory '/home/vmaffione/git/nxw/irati/build/librina/src'
make[2]: *** [Makefile:1518: install] Error 2
make[2]: Leaving directory '/home/vmaffione/git/nxw/irati/build/librina/src'
make[1]: *** [Makefile:476: install-recursive] Error 1
make[1]: Leaving directory '/home/vmaffione/git/nxw/irati/build/librina'
make: *** [Makefile:773: install] Error 2
install-user-from-scratch: Cannot complete librina phase

````

**Comment by vmaffione** [comment]
This may be useful
https://github.com/rakshasa/libtorrent/files/663302/0001-Fix-the-DH-parameters-generation-with-OpenSSL-1.1.patch.txt

**Comment by TRIA** [comment]
This same problem occurs with the latest (August) Raspberry Pi release, which is based on Debian Stretch.  Packages:
    apt list openssl  
    Listing... Done  
    openssl/stable,now 1.1.0f-3 armhf [installed,automatic]  

This version from the previous Raspberry Pi (Jessie?) release DOES work:
    apt list openssl  
    Listing... Done  
    openssl/oldstable,now 1.0.1t-1+deb8u6 armhf [installed,automatic]  

The big difference lies here.  From the NON-working one (first one):
`grep "struct dh_st" /usr/include/*/*.h
/usr/include/openssl/dh.h:/* typedef struct dh_st DH; */
/usr/include/openssl/evp.h:struct dh_st;
/usr/include/openssl/evp.h:int EVP_PKEY_set1_DH(EVP_PKEY *pkey, struct dh_st *key);
/usr/include/openssl/evp.h:struct dh_st *EVP_PKEY_get0_DH(EVP_PKEY *pkey);
/usr/include/openssl/evp.h:struct dh_st *EVP_PKEY_get1_DH(EVP_PKEY *pkey);
/usr/include/openssl/ossl_typ.h:typedef struct dh_st DH;`

From the working version (2nd one above):
`grep "struct dh_st" /usr/include/openssl/*.h
/usr/include/openssl/dh.h:/* typedef struct dh_st DH; */
/usr/include/openssl/dh.h:struct dh_st {
/usr/include/openssl/evp.h:        struct dh_st *dh;       /* DH */
/usr/include/openssl/evp.h:struct dh_st;
/usr/include/openssl/evp.h:int EVP_PKEY_set1_DH(EVP_PKEY *pkey, struct dh_st *key);
/usr/include/openssl/evp.h:struct dh_st *EVP_PKEY_get1_DH(EVP_PKEY *pkey);
/usr/include/openssl/ossl_typ.h:typedef struct dh_st DH;`

Note that the working one actually defines the structure in dh.h, the non-working one doesn't.

**Comment by edugrasa** [comment]
We need to check the library version and add pre-processor code (if) to write code for both versions (1.0 and 1.1). Howto:

* https://wiki.openssl.org/index.php/OpenSSL_1.1.0_Changes#Adding_forward-compatible_code_to_older_versions

**Comment by edugrasa** [comment]
Fixed by #1171
